src/topic.ts

```diff
@@ -797,6 +797,38 @@ export class Topic {
     return this.publisher.publishMessage(message, callback!);
   }
 
+  /**
+   * Publishes the message specified in the data buffer. The actual publish may be
+   * delayed if publisher-side flow control is enabled and set to Pause. Queued
+   * `publishWithFlowControl` calls will resolve in the order they're received, pausing
+   * again if space runs out due to subsequent resolves.
+   *
+   * The Promise's resolve() value will be another Promise<string> (in an array) that
+   * you can wait on for the message ID, from the message actually being sent.
+   *
+   * @param {Buffer} data The message to send
+   * @param {Attributes} [attrs] Attributes to attach to the message
+   * @returns {Promise<Array<Promise<string>>>} A Promise that resolves to an array
+   *   with a Promise that resolves to the message ID. In other words, when the first
+   *   Promise resolves, you can crack the array's shell to get a second Promise
+   *   that will resolve when the message was sent, and its ID is ready. This is
+   *   required because JavaScript runtimes will coalesce Promises.
+   *
+   * @example
+   * const messageIdPromises: Promise<string>[] = [];
+   * for (let i = 0; i < 1000; i++) {
+   *   const [idPromise] = await topic.publishWithFlowControl(messageBuffer);
+   *   messageIdPromises.push(idPromise);
+   * }
+   * const ids = await Promise.all(messageIdPromises);
+   */
+  publishWithFlowControl(
```

Comment:
I'm not sure `publishWithFlowControl` is the right name given they can also publish with the "error" flow control and that is still publishing with flow control. The coalescing Promises is kind of frustrating. I think it makes the method much more confusing. Confusing enough that it kind of makes me want to go back to the idea of having something like:

```
const messageIdPromises: Promise<string>[] = [];
for (let i = 0; i < 1000; i++) {
  await topic.readyForPublish(messageBuffer);
  const idPromise = topic.publish(messageBuffer);
  messageIdPromises.push(idPromise);
}
const ids = await Promise.all(messageIdPromises);
```

Maybe for Node, it doesn't make sense to have a "Pause" mode, just have "Ignore" (default) and "Error." If one is in "Error" mode, then one can avoid the error by waiting on the Promise returned by `readyForPublish`, which can guarantee that a slot is available. Thoughts?

Reply:
Yeah, I went back and forth on this a number of times. I agree this method is really confusing. :| But also, it's safer and easier from my experimentation to take care of the waiting down inside the library. I think I'd rather keep that structure, but I am totally open to new ideas about the method name and/or how to do the nested wait.

Reply:
I just have a really strong aversion to the introduction of Array. It feels like a hack. Given the choice between making the user do it by having a separate method or having to use the Array because of coalescing promises, I think the former is much cleaner.